### PR TITLE
Improve some tests

### PR DESCRIPTION
**Pull request by Bouncner**
Cleaning up my local branches ...

This PR add some tests to part that bothered me:
 * ART index: is a rather complex data structure with behaviours that should be tested properly (e.g., when nodes are full and need to be split)
* Pruning tests: more tests (especially near the bounds and non-existing values between gaps) + more data types to test
* Storage manager: it's a print test, so that was a clear thing to do for me. Further, it brought up the problem that has been fixed in #1067 by @Bensk1 .

### CODE REVIEW

File: src/test/statistics/chunk_statistics/min_max_filter_test.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+namespace opossum {
+
+template <typename T>
+class MinMaxFilterTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    _values = pmr_vector<T>{-1000, 2, 3, 4, 7, 8, 10, 17, 123456};
+    _min_value = *std::min_element(std::begin(_values), std::end(_values));
+    _max_value = *std::max_element(std::begin(_values), std::end(_values));
+    _in_between = static_cast<T>(_min_value + 0.5 * (_max_value - _min_value));
```
Comment (mrks) [inline src/test/statistics/chunk_statistics/min_max_filter_test.cpp]: ?

File: src/test/statistics/chunk_statistics/min_max_filter_test.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    _before_range = _min_value - 1;
```
Comment (mrks) [inline src/test/statistics/chunk_statistics/min_max_filter_test.cpp]: Could you doc these variable names, please? No idea what they mean.

File: src/test/statistics/chunk_statistics/min_max_filter_test.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+TYPED_TEST_CASE(MinMaxFilterTest, FilterTypes);
+
+TYPED_TEST(MinMaxFilterTest, CanPruneOnBounds) {
+  auto filter = std::make_unique<MinMaxFilter<TypeParam>>(this->_values.front(), this->_values.back());
+
+  for (const auto& value : this->_values) {
+    EXPECT_FALSE(filter->can_prune({value}, PredicateCondition::Equals));
+  }
+
+  EXPECT_TRUE(filter->can_prune({this->_before_range}, PredicateCondition::LessThan));
```
Comment (mrks) [inline src/test/statistics/chunk_statistics/min_max_filter_test.cpp]: A one-line comment in front of each of these blocks that tells us what it tests would be great.

File: src/test/storage/storage_manager_test.cpp [hunk 4]
```diff
@@ -115,4 +116,35 @@ TEST_F(StorageManagerTest, ListViewNames) {
...
 
+TEST_F(StorageManagerTest, Print) {
+  std::ostringstream output;
+  auto& sm = StorageManager::get();
+  sm.print(output);
+
+  auto output_string = output.str();
+
+  EXPECT_TRUE(output_string.find("===== Tables =====") != std::string::npos);
+  EXPECT_TRUE(output_string.find("==== table >> first_table << (0 columns, 0 rows in 0 chunks)") != std::string::npos);
```
Comment (mrks) [inline src/test/storage/storage_manager_test.cpp]: Mmmh, testing with an empty table doesn't do anything except for making the coverage misleadingly green. At least put in a simple table. Also, why are you `find`ing in the string instead of comparing it 1:1?

**Comment by Bouncner** [comment]
Ready to be reviewed again.

File: src/test/statistics/chunk_statistics/range_filter_test.cpp [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    for (auto it = value_set.begin(); it != std::prev(value_set.end()); ++it) {
+      auto begin = *it;
+      auto end = *(std::next(it));
+      begin_length_pairs.push_back(std::make_pair(begin, end - begin));
+    }
+
+    std::sort(begin_length_pairs.begin(), begin_length_pairs.end(),
+              [](auto& left, auto& right) { return left.second > right.second; });
+
+    for (int i = 0; i < gap_count && i < static_cast<int>(begin_length_pairs.size()); ++i) {
```
Comment (Bensk1) [inline src/test/statistics/chunk_statistics/range_filter_test.cpp]: Even though it is only a test, I believe we got rid of `i` and plain `int` in most places.

File: src/test/storage/adaptive_radix_tree_index_test.cpp [hunk 6]
```diff
@@ -43,6 +44,40 @@ class AdaptiveRadixTreeIndexTest : public BaseTest {
       pairs.emplace_back(std::make_pair(bc, values1[i]));
     }
     root = index1->_bulk_insert(pairs);
+
+    std::random_device rd;
+    _rng = std::mt19937(rd());
+  }
+
+  void _search_elements(std::vector<int>& values) {
```
Comment (Bensk1) [inline src/test/storage/adaptive_radix_tree_index_test.cpp]: More precise data types?

File: src/test/storage/adaptive_radix_tree_index_test.cpp [hunk 7]
```diff
@@ -140,7 +176,62 @@ TEST_F(AdaptiveRadixTreeIndexTest, VectorOfRandomInts) {
       expected_lower += 2;
     }
   }
-  EXPECT_EQ(index->upper_bound({99999}), index->cend());
+
+  for (int search_item = 0; search_item < static_cast<int>(3 * test_size); ++search_item) {
+    if (search_item % 2 == 0 && search_item < static_cast<int>(2 * test_size)) {
+      // all multiples of two within range of $ints should exist
```
Comment (Bensk1) [inline src/test/storage/adaptive_radix_tree_index_test.cpp]: `$`

File: src/test/storage/adaptive_radix_tree_index_test.cpp [hunk 8]
```diff
@@ -140,7 +176,62 @@ TEST_F(AdaptiveRadixTreeIndexTest, VectorOfRandomInts) {
       expected_lower += 2;
     }
   }
-  EXPECT_EQ(index->upper_bound({99999}), index->cend());
+
+  for (int search_item = 0; search_item < static_cast<int>(3 * test_size); ++search_item) {
```
Comment (Bensk1) [inline src/test/storage/adaptive_radix_tree_index_test.cpp]: I think a comment for the loop condition could be useful.

File: src/test/storage/adaptive_radix_tree_index_test.cpp [hunk 9]
```diff
@@ -140,7 +176,62 @@ TEST_F(AdaptiveRadixTreeIndexTest, VectorOfRandomInts) {
...
+
+    // search for elements not existing
+    EXPECT_EQ(index->lower_bound({search_item}), index->upper_bound({search_item}));
+  }
+}
+
+TEST_F(AdaptiveRadixTreeIndexTest, SimpleTest) {
+  std::vector<int> values = {0, 0, 0, 0, 0, 17, 17, 17, 99, std::numeric_limits<int>::max()};
+
+  auto segment = create_dict_segment_by_type<int>(DataType::Int, values);
```
Comment (Bensk1) [inline src/test/storage/adaptive_radix_tree_index_test.cpp]: Precise data types again

File: src/test/storage/adaptive_radix_tree_index_test.cpp [hunk 10]
```diff
@@ -140,7 +176,62 @@ TEST_F(AdaptiveRadixTreeIndexTest, VectorOfRandomInts) {
...
+  EXPECT_EQ(*index->lower_bound({17}), 5);
+  EXPECT_EQ(*index->upper_bound({17}), 8);
+  EXPECT_EQ(*index->lower_bound({99}), 8);
+  EXPECT_EQ(*index->upper_bound({99}), 9);
+  EXPECT_EQ(*index->lower_bound({std::numeric_limits<int>::max()}), 9);
+  EXPECT_EQ(index->upper_bound({std::numeric_limits<int>::max()}), index->cend());
+}
+
+/**
+* The following two tests try to test two rather extreme tests that both test the node overflow
```
Comment (Bensk1) [inline src/test/storage/adaptive_radix_tree_index_test.cpp]: Lots of `test` in that line, maybe `cases`?

File: src/test/storage/adaptive_radix_tree_index_test.cpp [hunk 11]
```diff
@@ -140,7 +176,62 @@ TEST_F(AdaptiveRadixTreeIndexTest, VectorOfRandomInts) {
...
+* handling of the ART implementation:
+*   - sparse vector: wide range of values (between 1 and MAX_INT) with large gaps
+*   - dense vector: expenential distribution, rounded to integer values
+**/
+TEST_F(AdaptiveRadixTreeIndexTest, SparseVectorOfRandomInts) {
+  // Increase temporaliy when working on the ART
```
Comment (Bensk1) [inline src/test/storage/adaptive_radix_tree_index_test.cpp]: `temporaliy` typo. What do you mean with the comment? Someone who is changing the implementation of the art should increase the test_size for the time working on it? Why? Because otherwise the test is not really helpful?

File: src/test/storage/adaptive_radix_tree_index_test.cpp [hunk 12]
```diff
@@ -140,7 +176,62 @@ TEST_F(AdaptiveRadixTreeIndexTest, VectorOfRandomInts) {
...
+  std::uniform_int_distribution<int> uni(1, std::numeric_limits<int>::max() - 1);
+
+  std::vector<int> values(test_size);
+  std::generate(values.begin(), values.end(), [this, &uni]() { return uni(_rng); });
+
+  _search_elements(values);
+}
+
+TEST_F(AdaptiveRadixTreeIndexTest, DenseVectorOfRandomInts) {
+  // Increase temporaliy when working on the ART
```
Comment (Bensk1) [inline src/test/storage/adaptive_radix_tree_index_test.cpp]: see above

File: src/test/storage/storage_manager_test.cpp [hunk 13]
```diff
@@ -115,4 +116,37 @@ TEST_F(StorageManagerTest, ListViewNames) {
...
+  sm.drop_table("first_table");
+  sm.drop_table("second_table");
+
+  sm.add_table("third_table", load_table("src/test/tables/int_float.tbl"));
+
+  sm.export_all_tables_as_csv(opossum::test_data_path);
+
+  const std::string filename = opossum::test_data_path + "/third_table.csv";
+  EXPECT_TRUE(filesystem::exists(filename));
+  std::remove(filename.c_str());
```
Comment (Bensk1) [inline src/test/storage/storage_manager_test.cpp]: There is also `std::filesystem::remove()`. I don't know if it is any better or worse, but it seems more consistent and does not rely on `char*` as input type.

File: src/test/storage/storage_manager_test.cpp [hunk 14]
```diff
@@ -115,4 +116,37 @@ TEST_F(StorageManagerTest, ListViewNames) {
...
+
+  EXPECT_TRUE(output_string.find("===== Views ======") != std::string::npos);
+  EXPECT_TRUE(output_string.find("==== view >> first_view <<") != std::string::npos);
+  EXPECT_TRUE(output_string.find("==== view >> second_view <<") != std::string::npos);
+}
+
+TEST_F(StorageManagerTest, ExportTables) {
+  std::ostringstream output;
+  auto& sm = StorageManager::get();
+  sm.drop_table("first_table");
```
Comment (Bensk1) [inline src/test/storage/storage_manager_test.cpp]: Seems a bit odd that an ExportTables test drops two tables. We should probably at least add a comment?

File: src/test/statistics/chunk_statistics/min_max_filter_test.cpp [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+template <>
+class MinMaxFilterTest<std::string> : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    _values = pmr_vector<std::string>{"aa", "bb", "b", "bbbbba", "bbbbbb", "bbbbbc", "c"};
+    _min_value = *std::min_element(std::begin(_values), std::end(_values));
+    _max_value = *std::max_element(std::begin(_values), std::end(_values));
+    _in_between = "ba";   // value in between the min and max
+    _before_range = "a";  // value smaller/before than the minimum
+    _after_range = "cc";  // value larger/beyond than the maximum
```
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/test/statistics/chunk_statistics/min_max_filter_test.cpp]: Does not need to be the successor (hard to determine for floating point). Should I add a comment?

File: src/test/statistics/chunk_statistics/range_filter_test.cpp [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+  void test_varying_range_filter_size(int gap_count) {
+    // RangeFilter constructor takes range count, not gap count
+    auto filter = RangeFilter<T>::build_filter(_values, gap_count + 1);
+
+    for (const auto& value : _values) {
+      EXPECT_FALSE(filter->can_prune({value}, PredicateCondition::Equals));
+    }
+
+    // find $gap_count largest gaps
+    auto value_set = std::set<T>(_values.begin(), _values.end());
```
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/test/statistics/chunk_statistics/range_filter_test.cpp]: Helps us only looking at distinct values and allowing to iterate in a sorted manner. Added a comment.

File: src/test/storage/adaptive_radix_tree_index_test.cpp (hunk 10 above)
Comment (Bouncner) [reply src/test/storage/adaptive_radix_tree_index_test.cpp]: Done.

File: src/test/storage/adaptive_radix_tree_index_test.cpp (hunk 11 above)
Comment (Bouncner) [reply src/test/storage/adaptive_radix_tree_index_test.cpp]: Old comment, I'll remove it.
The values should be sufficient (from my understanding) to cover the situations we want to test.

Actually, there is still more to test for the ART which is not covered well yet. But this PR is a starting point.

File: src/test/storage/storage_manager_test.cpp (hunk 14 above)
Comment (Bouncner) [reply src/test/storage/storage_manager_test.cpp]: Done.

File: src/test/statistics/chunk_statistics/min_max_filter_test.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (Bensk1) [reply src/test/statistics/chunk_statistics/min_max_filter_test.cpp]: No, it's fine. I just wanted to ask if you need the successor and whether this was correct.

File: src/test/statistics/chunk_statistics/range_filter_test.cpp [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+template <typename T>
+class RangeFilterTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    _values = pmr_vector<T>{-1000, 2, 3, 4, 7, 8, 10, 17, 100, 101, 102, 103, 123456};
+    _min_value = *std::min_element(std::begin(_values), std::end(_values));
+    _max_value = *std::max_element(std::begin(_values), std::end(_values));
+
+    // value in the largest gap of the test data
+    // we changing test data, ensure that value is not part of a range in ranges unless |ranges|==1)
```
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/test/statistics/chunk_statistics/range_filter_test.cpp]: Fixed.

File: src/test/statistics/chunk_statistics/min_max_filter_test.cpp [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+    _in_between = static_cast<T>(_min_value + 0.5 * (_max_value - _min_value));  // value in between the min and max
+    _before_range = _min_value - 1;                                              // value smaller than the minimum
+    _after_range = _max_value + 1;                                               // value larger than the maximum
+  }
+
+  pmr_vector<T> _values;
+  T _before_range, _min_value, _max_value, _after_range, _in_between;
+};
+
+// the test data for strings needs to handled differently from numerics
```
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/test/statistics/chunk_statistics/min_max_filter_test.cpp]: Done.

**Comment by Bouncner** [comment]
Thanks for the review. I think I've address all the issues brought up.

File: src/test/storage/adaptive_radix_tree_index_test.cpp (hunk 6 above)
Comment (Bouncner) [reply src/test/storage/adaptive_radix_tree_index_test.cpp]: Would make sense, but honestly I just extended the existing tests to cover more corner cases.
We should type these tests, but I doubt I find the time the upcoming weeks.

File: src/test/storage/adaptive_radix_tree_index_test.cpp (hunk 8 above)
Comment (Bouncner) [reply src/test/storage/adaptive_radix_tree_index_test.cpp]: Added a comment.
